### PR TITLE
Implemented unix command

### PR DESCRIPTION
**Pull request by likhithanimma1**
## Proposed changes

Implemented "Issue Unix Command" feature in Zowe Explorer

## Release Notes

Milestone:

Changelog:

## Types of changes

What types of changes does your code introduce to Zowe Explorer?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Updates to Documentation or Tests (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer_

- [ ] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/vscode-extension-for-zowe/wiki/Best-Practices:-Contributor-Guidance) wiki
- [x] PR title follows [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)
- [x] PR Description is included
- [ ] gif or screenshot is included if visual changes are made
- [x] `yarn workspace vscode-extension-for-zowe vscode:prepublish` has been executed
- [ ] All checks have passed (DCO, Jenkins and Code Coverage)
- [ ] I have added unit test and it is passing
- [ ] I have added integration test and it is passing
- [ ] There is coverage for the code that I have added
- [ ] I have tested it manually and there are no regressions found
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any PR dependencies have been merged and published (if appropriate)

### CODE REVIEW

**Comment by JillieBeanSim** [comment]
@likhithanimma1 shouldn't this point to the zowe:next branch?

**Comment by likhithanimma1** [comment]
> @likhithanimma1 shouldn't this point to the zowe:next branch?

Yes

**Review by adam-wolfe — DISMISSED** [review]

File: packages/zowe-explorer/src/command/UnixCommandHandler.ts [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+        }
+    }
+
+    public async setsshSession(): Promise<SshSession> {
+        ZoweLogger.trace("UnixCommandHandler.setsshSession called.");
+        const sshprofile: imperative.IProfileLoaded = Profiles.getInstance().getDefaultProfile("ssh");
+        if (!sshprofile) {
+            Gui.errorMessage(localize("setsshProfile.couldnotfindprofile", "Could not find the zowe SSH Profile.Create One"));
+        }
+        const cmdArgs: imperative.ICommandArguments = this.getCmdArgs(sshprofile?.profile as imperative.IProfileLoaded);
```
[the remark this answers is not in the document]
Comment (likhithanimma1) [reply packages/zowe-explorer/src/command/UnixCommandHandler.ts:160]: Yes Updated!

File: packages/zowe-explorer-api/src/extend/interfaces.ts [hunk 2]
```diff
@@ -470,4 +470,15 @@ export interface ICommand extends ICommon {
...
+    /**
+     * Issues a Unix Command and returns a Console Command API response.
+     *
+     * @param {string} command
+     * @param {string} cwd
+     * @returns {string>}
+     * @memberof ICommand
+     */
+    issueUnixCommand?(sshSession: zowe.SshSession, command: string, cwd: string): Promise<string>;
+    sshNeededforUnixCommand(): boolean;
```
[the remark this answers is not in the document]
Comment (likhithanimma1) [reply packages/zowe-explorer-api/src/extend/interfaces.ts]: Made it as optional.

File: packages/zowe-explorer-api/CHANGELOG.md [hunk 3]
```diff
@@ -172,3 +172,7 @@ All notable changes to the "zowe-explorer-api" extension will be documented in t
 ## `1.10.1`
 
 - Initial release
+
+## TBD Release
+
+- Implemented right click action for Issue UNIX Command and have it available via the Zowe Explorer API [#1326](https://github.com/zowe/vscode-extension-for-zowe/issues/1326)
```
[the remark this answers is not in the document]
Comment (likhithanimma1) [reply packages/zowe-explorer-api/CHANGELOG.md]: Addressed

File: packages/zowe-explorer/CHANGELOG.md [hunk 4]
```diff
@@ -926,3 +926,7 @@ All notable changes to the "vscode-extension-for-zowe" extension will be documen
 ## 0.5.0
 
 - Initial release
+
+## TBD Release
+
+- Implemented right click action for Issue UNIX Command and have it available via the Zowe Explorer API [#1326](https://github.com/zowe/vscode-extension-for-zowe/issues/1326)
```
[the remark this answers is not in the document]
Comment (likhithanimma1) [reply packages/zowe-explorer/CHANGELOG.md]: Addressed

File: packages/zowe-explorer/src/command/UnixCommandHandler.ts [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+                Gui.showMessage(localize("issueUnixCommand.noProfilesLoaded", "No profiles available"));
+                return;
+            }
+        } else {
+            profile = node.getProfile();
+        }
+        if (ZoweExplorerApiRegister.getCommandApi(profile).sshProfileRequired) {
+            this.sshSession = await this.setsshSession();
+        } else {
+            Gui.showMessage(localize("issueUnixCommand.notsupportedForProfile", "Action not being supported for the profile type ") + profile.type);
```
Comment (JillieBeanSim) [inline packages/zowe-explorer/src/command/UnixCommandHandler.ts:132]: lets make this string consistent with the others where the api doesn't exist for the extender's profile type, ie.
Gui.errorMessage(localize("issueMvsCommand.apiNonExisting", "Not implemented yet for profile of type: ") + profile.type); 
from MVS command.

**Review by JillieBeanSim — CHANGES_REQUESTED** [review]
this is looking really good, thanks @likhithanimma1 !  I did leave some comments on things I noticed while testing.
I would also like to mention that I got this error with a unix path of `/u/{userid}/projectFolder/copyTest (3)/` due to the spaces and parenthesis.  Will wrapping the command in quotes solve this so the action can be done? Also, when we output the command in output view should we also include the path the action is being done?  ie. `> {path} {command}` or other way around?

**Review by traeok — CHANGES_REQUESTED** [review]
This is coming along nicely @likhithanimma1!

I'm not sure if this is related to this PR, but I'm running into an issue where Zowe Explorer cannot find my default SSH profile when trying to test:

image

I briefly looked into where this could be coming from, and it turns out that `Profiles.getInstance().getDefaultProfile("ssh")` returns undefined in `setsshSession`:

image

File: packages/zowe-explorer/src/command/UnixCommandHandler.ts (hunk 5 above)
Comment (likhithanimma1) [reply packages/zowe-explorer/src/command/UnixCommandHandler.ts]: Added {path} to the output view. And wrapping the path in quotes helped to resolve the error. Modified the above Gui.errorMessage

File: packages/zowe-explorer/src/command/UnixCommandHandler.ts [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+                    return;
+                }
+            }
+        }
+        if (!session) {
+            const profiles = Profiles.getInstance();
+            const allProfiles: imperative.IProfileLoaded[] = profiles.allProfiles;
+            const profileNamesList = allProfiles.map((temprofile) => {
+                return temprofile.name;
+            });
```
[the remark this answers is not in the document]
Comment (likhithanimma1) [reply packages/zowe-explorer/src/command/UnixCommandHandler.ts]: Modified

File: packages/zowe-explorer/src/command/UnixCommandHandler.ts [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+                    placeHolder: localize("issueUnixCommand.quickPickOption", "Select the Profile to use to submit the Unix command"),
+                    ignoreFocusOut: true,
+                    canPickMany: false,
+                };
+                const sesName = await Gui.showQuickPick(profileNamesList, quickPickOptions);
+                if (sesName === undefined) {
+                    Gui.showMessage(localize("issueUnixCommand.cancelled", "Operation Cancelled"));
+                    return;
+                }
+                profile = allProfiles.filter((temprofile) => temprofile.name === sesName)[0];
```
[the remark this answers is not in the document]
Comment (likhithanimma1) [reply packages/zowe-explorer/src/command/UnixCommandHandler.ts]: Modified

**Review by JillieBeanSim — DISMISSED** [review]
This LGTM! thanks @likhithanimma1 for adding this awesome feature for v3 and thanks for addressing all our concerns and requests

**Review by JillieBeanSim — APPROVED** [review]

File: packages/zowe-explorer/CHANGELOG.md [hunk 8]
```diff
@@ -984,3 +984,7 @@ All notable changes to the "vscode-extension-for-zowe" extension will be documen
 ## 0.5.0
 
 - Initial release
+
+## TBD Release
+
+- Added the Issue UNIX Commands feature. [#1326](https://github.com/zowe/vscode-extension-for-zowe/issues/1326)
```
Comment (JillieBeanSim) [inline packages/zowe-explorer/CHANGELOG.md:990]: can we get this changelog moved to top of file please

**Review by JillieBeanSim — CHANGES_REQUESTED** [review]

**Review by traeok — CHANGES_REQUESTED** [review]
Thanks for addressing the missing `ssh` profile from the quick pick @likhithanimma1! I'm now seeing the `ssh` profiles in that list as expected.

However, after selecting my SSH profile and entering a path, an error dialog appears saying that Zowe Explorer tried to call a non-existing Common API:

image

**Review by t1m0thyj — CHANGES_REQUESTED** [review]
@likhithanimma1 Thanks for implementing this new feature 🙂 

A few comments:
* In the quickpick I would expect z/OSMF profiles to be excluded from the list:
    image
    I think we can use `ProfileManagement.getRegisteredProfileNameList` to filter the profile list, but his may depend on https://github.com/zowe/vscode-extension-for-zowe/pull/2539 which is being merged into "next" in https://github.com/zowe/vscode-extension-for-zowe/pull/2588.
* For the path prompt, can we make entering a path optional? If no path is entered, we could default to the user home directory (`~`) which matches the behavior of SSH.
    image

**Comment by zFernand0** [comment]
>In the quickpick I would expect z/OSMF profiles to be excluded from the list

I agree with that. Maybe we can add those back whenever we leverage the z/OSMF APIs for dealing with z/OS Unix commands

File: packages/zowe-explorer/CHANGELOG.md (hunk 8 above)
Comment (likhithanimma1) [reply packages/zowe-explorer/CHANGELOG.md]: Yes moved to the top

**Review by JillieBeanSim — APPROVED** [review]
LGTM! Thanks @likhithanimma1 for this new functionality and for addressing all of the feedback

**Comment by JillieBeanSim** [comment]
> All I see are zosmf profiles in the quick pick despite having a team config that contains ssh profiles: image If I select one of the zosmf profiles, I am still able to issue Unix commands.

The Profile seen in the list when going through the command palette are profiles registered with the UNIX files tree view.  The functionality here uses the default ssh profile, which is a good starting point and can be expanded in separate PRs to bring in capability to support a choice from multiple ssh profiles if available.

**Comment by traeok** [comment]
> The Profile seen in the list when going through the command palette are profiles registered with the UNIX files tree view that the action is driven from. The functionality here uses the default ssh profile, which is a good starting point and can be expanded in separate PRs to bring in capability to support a choice from multiple ssh profiles if available.

I think it makes sense to offer the profile selection as a future enhancement - since the logic will pick the default SSH profile regardless of what profile is chosen in this quick pick, can we hide this quick pick until profile selection is supported?

**Review by traeok — APPROVED** [review]
@JillieBeanSim Thank you for clarifying on extenders, I did not fully realize the scenario for extenders that do not require SSH for Unix commands.

If the fallback to the default SSH profile is planned to stay in-place, we might want to mention this behavior in Zowe Docs for awareness - otherwise users might try to execute Unix commands through a z/OSMF profile without having any default SSH profile in their config. 

_@anaxceron tagging for reference regarding v3 readiness/feature doc 😅_

---

Thanks @likhithanimma1 for this feature & addressing the feedback! Tested and working as expected, this is a big WOW for v3 😋 LGTM

**Pull request merged**